### ISSUE TITLE
Variable stiffness and damping about the soft objects

### ISSUE THREAD

**Issue by JadeCong** [hermanjakobsen/robotic-ultrasound-imaging#7]
Hello，you did a great job about soft object manipulation. Then I have a question about the soft object: How do you deal with the variable stiffness and damping of soft objects? The robot can adjust the contact force as desired or not if the stiffness and damping changed at different areas of the soft object when scanning. I really want to know how do you figure it out. Appreciate if I get a reply.

**Comment by hermanjakobsen** [comment]
Hey! 

I did not fully understand the question, so I will try my best to give an overview of my thoughts :) 

According to the MuJoCo [documentation](http://www.mujoco.org/book/modeling.html#CComposite), soft objects are composed of rigid objects placed in a grid. Each rigid body has a sliding joint connected to the center, which allows the surface of the soft object to compress and expand at any point. 

Due to how soft objects are modelled, there are, to my understanding, several parameters deciding the stiffness and damping of the soft objects. Two parameters are the stiffness and damping of the joints connecting the rigid bodies to the center. Other parameters are the size, spacing, shape and mass of the rigid bodies themselves. In order to introduce some stiffness and damping variations of the soft object when training the RL agent, I randomize the joint stiffness and damping between each episode. 

The task of the robot is to follow a trajectory along the surface of a soft object (i.e. a sweep), while exerting a desired contact force. The robot uses the operational space controller to track the trajectory in the xy-plane, while having a compliant behaviour in the z-direction. This makes it possible for the robot to move freely in the z-direction while sweeping, making the robot able to compensate for movement and uncertainty of the soft object. The RL agent is set to learn the gains of the OSC.

A more thorough description of the project is presented in my master's thesis, and it will hopefully be publicly available soon! :) 

Hope this clarified things a bit!

**Comment by JadeCong** [comment]
> Hey!
> 
> I did not fully understand the question, so I will try my best to give an overview of my thoughts :)
> 
> According to the MuJoCo [documentation](http://www.mujoco.org/book/modeling.html#CComposite), soft objects are composed of rigid objects placed in a grid. Each rigid body has a sliding joint connected to the center, which allows the surface of the soft object to compress and expand at any point.
> 
> Due to how soft objects are modelled, there are, to my understanding, several parameters deciding the stiffness and damping of the soft objects. Two parameters are the stiffness and damping of the joints connecting the rigid bodies to the center. Other parameters are the size, spacing, shape and mass of the rigid bodies themselves. In order to introduce some stiffness and damping variations of the soft object when training the RL agent, I randomize the joint stiffness and damping between each episode.
> 
> The task of the robot is to follow a trajectory along the surface of a soft object (i.e. a sweep), while exerting a desired contact force. The robot uses the operational space controller to track the trajectory in the xy-plane, while having a compliant behaviour in the z-direction. This makes it possible for the robot to move freely in the z-direction while sweeping, making the robot able to compensate for movement and uncertainty of the soft object. The RL agent is set to learn the gains of the OSC.
> 
> A more thorough description of the project is presented in my master's thesis, and it will hopefully be publicly available soon! :)
> 
> Hope this clarified things a bit!

**Comment by JadeCong** [comment]
Yeah, it's pretty clear, and thank you so much. I wanna read your thesis as soon as possible, but here now I am a little confused about other situations as following:
1. According to your description(track the trajectory in the xy-plane, while having a compliant behavior in the z-direction) and running demo from your project, as far as I understand, the robot can only adjust the contact force between the probe and soft object in z direction which is perpendicular to desktop plane with regard to env coordinate system? Is that right?
2. I am curious about whether the learned policy could adjust the pose(include orientation) of ultrasound probe to meet the contents that the axis of probe is perpendicular to curved surface of soft object with desired const contact force in real time. This  question seems like the continuance of first question. haha~ :)
3. Well, while runing the demo, I found the learned policy control will end if the probe lose contact with soft torso in the _check_terminated() function, and I must keep the probe touching the soft torso at the beginning of running policy according the _get_initial_qpos() funtion. Is that right? So how could I get the policy resumes running at terminated pose(losing touch with torso) and keep scanning with rendering scene instead of reinitializing the beginning pose?
4. Could I modify the waypoints of the reference trajectory with custom function that make each waypoint contains pose of which z direction is perpendicular to torso surface, and then run the policy successfully?

Again, you did an amazing job. I am also working on soft object manipulation, and I really appreciate your detailed answers. After then I may get some idea about uncertainty from other view.

**Comment by JadeCong** [comment]
Yeah, it's pretty clear, and thank you so much. I wanna read your thesis as soon as possible, but here now I am a little confused about other situations as following:

According to your description(track the trajectory in the xy-plane, while having a compliant behavior in the z-direction) and running demo from your project, as far as I understand, the robot can only adjust the contact force between the probe and soft object in z direction which is perpendicular to desktop plane with regard to env coordinate system? Is that right?
I am curious about whether the learned policy could adjust the pose(include orientation) of ultrasound probe to meet the contents that the axis of probe is perpendicular to curved surface of soft object with desired const contact force in real time. This question seems like the continuance of first question. haha~ :)
Well, while runing the demo, I found the learned policy control will end if the probe lose contact with soft torso in the _check_terminated() function, and I must keep the probe touching the soft torso at the beginning of running policy according the _get_initial_qpos() funtion. Is that right? So how could I get the policy resumes running at terminated pose(losing touch with torso) and keep scanning with rendering scene instead of reinitializing the beginning pose?
Could I modify the waypoints of the reference trajectory with custom function that make each waypoint contains pose of which z direction is perpendicular to torso surface, and then run the policy successfully?
Again, you did an amazing job. I am also working on soft object manipulation, and I really appreciate your detailed answers. After then I may get some idea about torso's uncertainty from other view.

**Comment by JadeCong** [comment]
Yeah, it's pretty clear, and thank you so much. I wanna read your thesis as soon as possible, but here now I am a little confused about other situations as following:
1. According to your description(track the trajectory in the xy-plane, while having a compliant behavior in the z-direction) and running demo from your project, as far as I understand, the robot can only adjust the contact force between the probe and soft object in z direction which is perpendicular to desktop plane with regard to env coordinate system? Is that right?
2. I am curious about whether the learned policy could adjust the pose(include orientation) of ultrasound probe to meet the contents that the axis of probe is perpendicular to curved surface of soft object with desired const contact force in real time. This question seems like the continuance of first question. haha~ :)
3. Well, while runing the demo, I found the learned policy control will end if the probe lose contact with soft torso in the _check_terminated() function, and I must keep the probe touching the soft torso at the beginning of running policy according the _get_initial_qpos() funtion. Is that right? So how could I get the policy resumes running at terminated pose(losing touch with torso) and keep scanning with rendering scene instead of reinitializing the beginning pose?
4. Could I modify the waypoints of the reference trajectory with custom function that make each waypoint contains pose of which z direction is perpendicular to torso surface, and then run the policy successfully?

Again, you did an amazing job. I am also working on soft object manipulation, and I really appreciate your detailed answers. After then I may get some idea about uncertainty from other view.

**Comment by JadeCong** [comment]
Yeah, it's pretty clear, and thank you so much. I wanna read your thesis as soon as possible, but here now I am a little confused about other situations as following:
1. According to your description(track the trajectory in the xy-plane, while having a compliant behavior in the z-direction) and running demo from your project, as far as I understand, the robot can only adjust the contact force between the probe and soft object in z direction which is perpendicular to desktop plane with regard to env coordinate system? Is that right?
2. I am curious about whether the learned policy could adjust the pose(include orientation) of ultrasound probe to meet the contents that the axis of probe is perpendicular to curved surface of soft object with desired const contact force in real time. This question seems like the continuance of first question. haha~ :)
3. Well, while runing the demo, I found the learned policy control will end if the probe lose contact with soft torso in the _check_terminated() function, and I must keep the probe touching the soft torso at the beginning of running policy according the _get_initial_qpos() funtion. Is that right? So how could I get the policy resumes running at terminated pose(losing touch with torso) and keep scanning with rendering scene instead of reinitializing the beginning pose?
4. Could I modify the waypoints of the reference trajectory with custom function that make each waypoint contains pose of which z direction is perpendicular to torso surface, and then run the policy successfully?

Again, you did an amazing job. I am also working on soft object manipulation, and I really appreciate your detailed answers. After then I may get some idea about torso's uncertainty from other view.

**Comment by hermanjakobsen** [comment]
Great questions!

1. Yes! The agent is given a reward for adjusting the contact force in the z-direction by moving the the probe freely in the z-direction. Ideally, the robot should be able to adjust the contact force normal between the soft object and the probe, but unfortunately I did not have time to implement it :/. It is possible to extract the contact force normal in MuJoCo, so it should not be too much work to implement this feature. 
2. Yes, I think that should be possible. As of now, the `goal_quat` variable is just set to a constant value corresponding to the probe being in an upright position. Using information and measurements from the contact at each timestep, it should be possible to calculate an orientation of the probe that satisfies your criteria. The `goal_quat` variable can then be updated with this calculated value at each timestep, and the agent will then be given reward on how well it "tracks" this desired orientation. Hopefully, the agent will learn how to orientate the probe based on the surface curvature. 
3. If the probe loses contact with the soft torso, after initially having made contact with the torso, the episode will end. The `_get_inital_qpos()` is made such that the episode is initialized with the probe touching the torso. All the early termination criteria can be deactivated with by setting the `early_termination` flag when creating the environment. By setting the flag to `False`, episode will not end if the probe loses contact. 
4. Yes, it is possible to fully customize your reference trajectory. The desired position can be implemented by updating `traj_pt` at each timestep, and the desired orientation can be implemented by updating `goal_quat` at each timestep. Note that the agent is only given reward for tracking the x- and y-position and the orientation, but this can easily be modified by changing the reward function.

**Comment by JadeCong** [comment]
Well, I customized the reference trajectories, and then I will get the normal contact force to make the agent learn orientation adjustment. Thanks~
![custom_trajectory](https://user-images.githubusercontent.com/16516205/126923797-4432bf6d-c5f7-4b42-b546-925b344eea8e.png)
